lib/source/defender/aws_iot_defender_api.c

```diff
@@ -551,18 +589,105 @@ void _rejectCallback( void * pArgument,
     AwsIotDefender_Assert( pPublish->u.message.info.pPayload );
 
     /* Invoke user's callback with rejected event. */
+    _handleApplicationCallback( AWS_IOT_DEFENDER_METRICS_REJECTED, pPublish );
+    
+}
+
+/*-----------------------------------------------------------*/
+
+void _mqttAsyncOperationCallback( void * pContext,
+                                  IotMqttCallbackParam_t * pCallbackParams )
+{
+    IotLogInfo( "MQTT Async operation callback" );
+    IotMqttError_t mqttError = pCallbackParams->u.operation.result;
+    IotMqttOperationType_t mqttOp = pCallbackParams->u.operation.type;
+    IotTaskPoolError_t taskPoolError = IOT_TASKPOOL_SUCCESS;
+
+    IotLogDebug( "Operation = %d status = %d\n", mqttOp, mqttError );
+
+    if( mqttError != IOT_MQTT_SUCCESS )
+    {
+        IotLogError( "Operation = %d Failed status = %d\n", mqttOp, mqttError );
+        /* Async operation has failed. Notify application and terminate the defender state machine */
+        _handleApplicationCallback( AWS_IOT_DEFENDER_FAILURE_MQTT, NULL );
+    }
+    else
+    {
+        switch( mqttOp )
+        {
+            case IOT_MQTT_SUBSCRIBE:
+                _subscribed = true;
+                /* Create metrics subscribe job. */
+                taskPoolError = IotTaskPool_CreateJob( _metricsPublishRoutine, NULL, &_metricsPublishJobStorage, &_metricsPublishJob );
```

Comment:
This is not safe.  MQTT is a network protocol, you could receive a duplicated response.  That would cause this callback to be invoked more than once, with both invocations modifying this global data.